gdb-server/src/stub.rs

```diff
@@ -0,0 +1,132 @@
+    ///                       Multiple instances are bound by adding an offset to the supplied port.
+    ///
+    /// # Returns
+    /// Vec with the computed configuration
+    pub fn from_session(
+        session: &Session,
+        connection_string: Option<impl Into<String>>,
+    ) -> Vec<Self> {
+        let connection_string = connection_string
+            .map(|cs| cs.into())
+            .unwrap_or_else(|| CONNECTION_STRING.to_owned());
+
+        let addrs: Vec<SocketAddr> = connection_string.to_socket_addrs().unwrap().collect();
+
+        // Build a grouped list of cores by core type
+        // GDB only supports one architecture per stub so if we have two core types,
+        // such as ARMv7-a + ARMv7-m, we must create two stubs to connect to.
+        let groups = session
+            .target()
+            .cores
+            .iter()
+            .enumerate()
+            .map(|(i, core)| (core.core_type, i))
+            .into_group_map();
+
+        // Create a GDB instance for each group, starting at the specified connection and adding one to the port each time
+        // For example - consider two groups computed above and an input of localhost:1337.
+        // Group 1 will bind to localhost:1337
+        // Group 2 will bind to localhost:1338
+        let ret = groups
+            .iter()
+            .enumerate()
+            .map(|(i, (core_type, cores))| GdbInstanceConfiguration {
+                core_type: *core_type,
+                cores: cores.to_vec(),
+                socket_addrs: adjust_addrs(&addrs, i),
+            })
+            .collect();
+
+        ret
+    }
+}
+
+/// Run a new GDB session.
+///
+/// # Arguments
+///
+/// * session - The [Session] to use
+/// * instances - a list of [GdbInstanceConfiguration] objects used to configure the GDB session
+///
+/// # Remarks
+///
+/// A default configuration can be created by calling [GdbInstanceConfiguration::from_session()]
+pub fn run<'a>(
+    session: Session,
+    instances: impl Iterator<Item = &'a GdbInstanceConfiguration>,
+) -> Result<()> {
+    // We can't use &mut Session because we have more than one target active.
+    // Rc / RefCell give the same behavior as &mut with runtime borrow checking.
+    // Arc / Mutex is not needed because all logic in this stub runs on one thread.
+    let session = Rc::new(RefCell::new(session));
+
+    // Turn our group list into GDB targets
+    let mut targets = instances
+        .map(|instance| {
+            target::RuntimeTarget::new(
+                session.clone(),
+                instance.cores.to_vec(),
+                &instance.socket_addrs[..],
+            )
+        })
+        .collect::<Result<Vec<target::RuntimeTarget>, Error>>()?;
+
+    // Process every target in a loop
+    loop {
+        let mut wait_time = Duration::ZERO;
+
+        for target in targets.iter_mut() {
+            wait_time = wait_time.min(target.process()?);
```

Comment:
Can this `min` ever produce something other than `Duration::ZERO`?

Reply:
Maybe it should have been a max? Not sure. I would agree that the min is useless if `process()` returns strictly positive numbers.